### PR TITLE
Modernize integration tests of TraceGraph_ELBO

### PR DESCRIPTION
**Pull request by fritzo**
This updates integration tests of `TraceGraph_ELBO` to:
- test reparameterized `Gamma` and `Beta` if available
- vectorize some of the `observe` statements to speed up tests
  
## Tested
- ran locally against PyTorch 0.3

### CODE REVIEW

File: tests/integration_tests/test_tracegraph_elbo.py [hunk 1]
```diff
@@ -230,29 +224,30 @@ def setUp(self):
...
-        self.beta_n = self.beta0 - data_sum + \
-            Variable(torch.Tensor([self.n_data]))
-        # posterior beta
+        self.beta_n = self.beta0 - data_sum + Variable(torch.Tensor([self.n_data]))  # posterior beta
         self.log_alpha_n = torch.log(self.alpha_n)
         self.log_beta_n = torch.log(self.beta_n)
 
+    @pytest.mark.skipif(not dist.beta.reparameterized, reason='not implemented')
+    def test_elbo_reparameterized(self):
+        self.do_elbo_test(True, 3000, 0.95, 0.0007)
```
Comment (martinjankowiak) [inline tests/integration_tests/test_tracegraph_elbo.py]: can't you do fewer steps, lower momentum, and higher lr here?

File: tests/integration_tests/test_tracegraph_elbo.py [hunk 2]
```diff
@@ -230,29 +224,30 @@ def setUp(self):
...
         pyro.clear_param_store()
+        beta = dist.beta if reparameterized else fakes.nonreparameterized_beta
 
         def model():
-            p_latent = pyro.sample("p_latent", dist.beta, self.alpha0, self.beta0)
-            for i, x in enumerate(self.data):
-                pyro.observe("obs_{}".format(i), dist.bernoulli, x,
-                             torch.pow(torch.pow(p_latent, 2.0), 0.5))
+            p_latent = pyro.sample("p_latent", beta, self.alpha0, self.beta0)
+            pyro.sample("obs", dist.bernoulli, torch.pow(torch.pow(p_latent, 2.0), 0.5),
```
Comment (martinjankowiak) [inline tests/integration_tests/test_tracegraph_elbo.py]: can remove this double power... this is holder from attempts to do fancier dependency tracking...

File: tests/integration_tests/test_tracegraph_elbo.py [hunk 3]
```diff
@@ -288,26 +281,28 @@ def setUp(self):
...
         self.alpha_n = self.alpha0 + sum_data  # posterior alpha
-        self.beta_n = self.beta0 + \
-            Variable(torch.Tensor([self.n_data]))  # posterior beta
+        self.beta_n = self.beta0 + len(self.data)  # posterior beta
         self.log_alpha_n = torch.log(self.alpha_n)
         self.log_beta_n = torch.log(self.beta_n)
 
+    @pytest.mark.skipif(not dist.gamma.reparameterized, reason='not implemented')
+    def test_elbo_reparameterized(self):
+        self.do_elbo_test(True, 8000, 0.95, 0.0007)
```
Comment (martinjankowiak) [inline tests/integration_tests/test_tracegraph_elbo.py]: same here re: n_steps

File: tests/integration_tests/test_tracegraph_elbo.py [hunk 4]
```diff
@@ -349,21 +344,27 @@ def setUp(self):
...
-        self.beta_n = self.beta0 + torch.sum(self.data)  # posterior beta
+        self.data = Variable(torch.Tensor([[3.0], [2.0]]))  # two observations
+        self.alpha_n = self.alpha0 + self.n_data  # posterior alpha
+        self.beta_n = self.beta0 + self.data.sum(0)  # posterior beta
         self.log_alpha_n = torch.log(self.alpha_n)
         self.log_beta_n = torch.log(self.beta_n)
 
+    @pytest.mark.skipif(not dist.gamma.reparameterized, reason='not implemented')
+    def test_elbo_reparameterized(self):
+        self.do_elbo_test(True, 8000, 0.95, 0.0007)
```
Comment (martinjankowiak) [inline tests/integration_tests/test_tracegraph_elbo.py]: and here?

File: tests/integration_tests/test_tracegraph_elbo.py (hunk 1 above)
Comment (fritzo) [reply tests/integration_tests/test_tracegraph_elbo.py]: I was having trouble getting tests to pass with reparameterized gradients. I was hoping to merge this big PR and then follow up by focusing on getting the tests to pass.

File: tests/integration_tests/test_tracegraph_elbo.py (hunk 2 above)
Comment (fritzo) [reply tests/integration_tests/test_tracegraph_elbo.py]: Done.

File: tests/integration_tests/test_tracegraph_elbo.py (hunk 3 above)
Comment (fritzo) [reply tests/integration_tests/test_tracegraph_elbo.py]: ditto: I'd like tune parameters in a follow-up PR

File: tests/integration_tests/test_tracegraph_elbo.py (hunk 4 above)
Comment (fritzo) [reply tests/integration_tests/test_tracegraph_elbo.py]: ditto: I'd like tune parameters in a follow-up PR
